gallery_dl/extractor/generic.py

```diff
@@ -0,0 +1,207 @@
+        will be matched by the "imageurl_pattern_ext" pattern below.
+        """
+        imageurl_pattern_src = r"""(?ix)
+            <(?:img|video|source)\s.*?      # <img>, <video> or <source>
+            src(?:set)?=["']?               # src or srcset attributes
+            (?P<URL>[^"'\s>]+)              # url
+            """
+
+        """
+        2: Look anywhere for urls containing common image/video extensions
+
+        The list of allowed extensions is borrowed from the directlink.py
+        extractor; other could be added, see
+        https://en.wikipedia.org/wiki/List_of_file_formats
+
+        We don't allow relative urls here, i.e. generic strings containing an
+        image filename (such as in <... alt="image.jpg">) because they often
+        get resolved into non-existing absolute urls, leading to 404 errors
+        that might confuse the user.
+
+        Compared to the "pattern" class variable, here we must exclude also
+        other special characters (space, ", ', >), since we are looking for
+        urls in html tags.
+
+        Note that we use the x flag, so spaces and "#" must be
+        backslash-quoted, unless they are in a character class ([]).
+        """
+        imageurl_pattern_ext = r"""(?ix)
+            (?:https?://[^/?&#"'>\s]+)          # scheme + domain
+            (?:[^?&#"'>\s]+)                    # path until dot+ext
+            \.(?:jpe?g|jpe|png|gif
+                 |web[mp]|mp4|mkv|og[gmv]|opus) # dot + image/video extensions
+            (?:\?[^"'>\s]*)?                    # optional query and fragment
+            """
+
+        imageurls_src = re.findall(imageurl_pattern_src, page)
+        imageurls_ext = re.findall(imageurl_pattern_ext, page)
+        imageurls = imageurls_src + imageurls_ext
+
+        # Resolve relative urls
+        #
+        # Some of the urls may be relative, so we resolve them prefixing them
+        # either by the page root url (self.root) if the relative url starts
+        # with "/", or by a proper "base" url if the relative url doesn't start
+        # with "/"
+
+        # If the page contains a <base> element, use it as base url
+        basematch = re.search(
+            r"(?i)(?:<base\s.*?href=[\"']?)(?P<url>[^\"' >]+)", page)
+        if basematch:
+            self.baseurl = basematch.group('url')
+        else:
+            # Use self.url if it doesn't end with an extension
+            if os.path.splitext(self.url)[1] == "":
+                self.baseurl = self.url
+            # Otherwise, strip the last part
+            else:
+                self.baseurl = os.path.dirname(self.url)
+
+        # Really resolve relative urls
+        absimageurls = []
+        for u in imageurls:
+            if u.startswith('http'):
+                absimageurls.append(u)
+            elif u.startswith('/'):
+                absimageurls.append(self.root + '/' + u)
+            else:
+                absimageurls.append(self.baseurl + '/' + u)
```

Comment:
This doesn't properly work on Wikipedia, for example.
https://en.wikipedia.org/wiki/Geometric_Shapes

I think you can just use `text.urljoin(u, self.baseurl)` here.

Reply:
Interesting, the problem here is that I wasn't supporting image urls starting with `//` (which are valid urls, requiring the client to reuse the same scheme used in the "previous" request); I've now added this case, completing `//` urls with `self.scheme` (read from the page url pattern, or defaulting to https).

BTW, before that, I tried passing the `//` url as is to the directlink extractor, but I got the message `URLs are not supported/enabled` from `DownloadJob.get_downloader()`, so maybe we might want to add support for `//` urls also there, in case some other extractors produce this kind of image urls.